internal/topology/variables/cluster_variable_validation.go

```diff
@@ -21,87 +21,107 @@ import (
+	if validateRequired {
+		allErrs = append(allErrs, validateRequiredVariables(valuesMap, defIndex, fldPath)...)
 	}
-	return allErrs
-}
-
-// validateRequiredClusterVariables validates all required variables from the ClusterClass exist in the Cluster.
-func validateRequiredClusterVariables(clusterVariables map[string]*clusterv1.ClusterVariable, clusterClassVariables map[string]*clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
-	var allErrs field.ErrorList
 
-	for variableName, clusterClassVariable := range clusterClassVariables {
-		if !clusterClassVariable.Required {
+	for i, value := range values {
+		// Values must have an associated definition and must have a non-empty definitionFrom if there are conflicting definitions.
+		definition, err := defIndex.get(value.Name, value.DefinitionFrom)
+		if err != nil {
+			allErrs = append(allErrs, field.Required(fldPath.Index(i), err.Error())) // TODO: consider if to add ClusterClass name
 			continue
 		}
 
-		if _, ok := clusterVariables[variableName]; !ok {
-			allErrs = append(allErrs, field.Required(fldPath,
-				fmt.Sprintf("required variable with name %q must be defined", variableName))) // TODO: consider if to use "Clusters with ClusterClass %q must have a variable with name %q"
-		}
+		// Values must be valid according to the schema in their definition.
+		allErrs = append(allErrs, ValidateClusterVariable(value.DeepCopy(), &clusterv1.ClusterClassVariable{
+			Name:     value.Name,
+			Required: definition.Required,
+			Schema:   definition.Schema,
+		}, fldPath.Index(i))...)
 	}
 
 	return allErrs
 }
 
-// validateClusterVariablesDefined validates all variables from the Cluster are defined in the ClusterClass.
-func validateClusterVariablesDefined(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables map[string]*clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+// validateRequiredVariables validates all required variables from the ClusterClass exist in the Cluster.
+func validateRequiredVariables(values map[string]map[string]clusterv1.ClusterVariable, definitions definitionsIndex, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
 
-	for i, clusterVariable := range clusterVariables {
-		if _, ok := clusterClassVariables[clusterVariable.Name]; !ok {
-			return field.ErrorList{field.Invalid(fldPath.Index(i).Child("name"), clusterVariable.Name,
-				fmt.Sprintf("variable with name %q is not defined in ClusterClass `status.variables`", clusterVariable.Name))} // TODO: consider if to add ClusterClass name
+	for name, perPatchDefs := range definitions {
+		for _, def := range perPatchDefs {
+			// Check the required value for the specific variable definition. If the variable is not required continue.
+			if !def.Required {
+				continue
+			}
+
+			// If there is no variable with this name defined in the Cluster add an error and continue.
+			valuesForName, found := values[name]
```

Comment:
Is this valuesByDefinitionFrom?

Reply:
This is all variable values with a given name

Reply:
Ah I see, yeah makes sense to think this way around :)